.github/workflows/ci.yml

```diff
@@ -0,0 +1,80 @@
+name: CI
+on:
+  push:
+    branches:
+      - master
+    tags:
+      - v[0-9]+.[0-9]+.[0-9]+*
+  pull_request:
+    branches:
+      - master
+jobs:
+ build:
+  name: Build
+  runs-on: ubuntu-latest
+
+  steps:
+  - uses: actions/checkout@v3
+  - name: Set up Ruby
+    uses: ruby/setup-ruby@v1
+    with:
+      ruby-version:  3.1
+  - name: Install dependencies
+    run: bundle install
+  - name: Build
+    run: gem build *.gemspec
+  - name: 'Upload Artifact'
+    uses: actions/upload-artifact@v3
+    with:
+      name: gems
+      path: '*.gem'
+ test:
+  runs-on: ubuntu-latest
+  needs: build
+  strategy:
+   matrix:
+    ruby-version: ['2.6', '2.7', '3.0', '3.1']
+
+  steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Set up Ruby latest
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version: 3.1
+    - name: Install dependencies
+      run: bundle install
+    - name: Run tests and collect coverage
+      run: bundle exec rake
+    - name: Upload coverage to Codecov
+      if: ${{ matrix.ruby-version == 3.1 }}
+      uses: codecov/codecov-action@v3
+      with:
+        files: ${{ github.workspace }}/coverage/coverage.xml
+ publish:      
+    name: Publish
+    if: startsWith(github.ref, 'refs/tags/v')
+    needs: [build, test]
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - run: sudo apt-get install -y oathtool
+    - name: Download all workflow run artifacts
+      uses: actions/download-artifact@v3
+      with:
+        name: gems
+        path: gems
+    - name: Set up Ruby
+      uses: ruby/setup-ruby@v1
+      with:
+        ruby-version:  3.1
+    - name: Publish gems to Rubygems
+      run: |
+       mkdir -p $HOME/.gem
+       touch $HOME/.gem/credentials
+       chmod 0600 $HOME/.gem/credentials
+       printf -- "---\n:rubygems_api_key: ${GEM_API_KEY}\n" > $HOME/.gem/credentials
+       gem push gems/*.gem --otp $(oathtool --base32 --totp "$GEM_2FA_SECRET_KEY")
```

Comment:
Can you explain this or share some resources? What is happening here

Reply:
Sure, you can check this post 
https://andrewm.codes/blog/automating-ruby-gem-releases-with-github-actions/
and line no 77 i have added 2FA auth, although i am not sure 2fa is mandatory or not in razorpay repo [rubygems](https://rubygems.org/gems/razorpay) until i added as a collaborator. but if there is no 2FA mandatory, then we will remove `--otp $(oathtool --base32 --totp "$GEM_2FA_SECRET_KEY")`

created clone of this repo and publish it (for testing). Check this once 
https://github.com/ankitdas13/razorruby/actions/runs/4805861071

Reply:
on the shared link it says we are building the image and publishing it. My concern is we have a build step do we need to build again? Can't we directly publish it?

Reply:
if we want to publish it directly then we have to save the build with help of this `actions/upload-artifact@v3` in git environment and while publish we can use this save build by using this `actions/download-artifact@v3`